src/selector.ts

```diff
@@ -0,0 +1,298 @@
+    for (let token of tokens) {
+      this.props.push(new OpProp(token));
+    }
+  }
+}
+
+/**
+Brackets holds parsed info for selections like `foo[(prop1=23, prop2)]`
+*/
+class Brackets {
+  op: Op; // the operation from `foo[!(operation)]`
+  prop: string; // `foo` from `foo[...]`
+  constructor(public raw: string) {
+    raw = raw.trim();
+    if (raw.includes('[') === false) throw new Error('Could not parse brackets: ' + raw);
+    if (raw.endsWith(']') === false) throw new Error('Could not parse brackets: ' + raw);
+    let tokens = raw.split('[');
+    if (tokens.length != 2) throw new Error('Could not parse brackets: ' + raw);
+    this.prop = tokens[0];
+    this.op = new Op(tokens[1].substring(0, tokens[1].length - 1));
+  }
+}
+
+class PathElement {
+  kind: ElementKind;
+  brackets?: Brackets;
+  parsedInfo: { [key: string]: any }; // per-kind info populated in `parse` calls
+
+  // This throws an exception if `raw` can not be parsed
+  constructor(public raw: string) {
+    this.kind = PathElement.sniffKind(raw);
+    this.parsedInfo = {};
+    this.parse();
+  }
+
+  select(target: any): any | undefined {
+    switch (this.kind) {
+      case ElementKind.Pluck:
+        return this.selectPluck(target);
+      case ElementKind.Index:
+        return this.selectIndex(target);
+      case ElementKind.Pick:
+        return this.selectPick(target);
+        break;
+      case ElementKind.Omit:
+        throw new Error('TBD');
+        break;
+      case ElementKind.Prefix:
+        throw new Error('TBD');
+        break;
+      case ElementKind.Suffix:
+        throw new Error('TBD');
+        break;
+      case ElementKind.Filter:
+        throw new Error('TBD');
+        break;
+       default:
+         throw new Error('Unknown PathElement.kind: ' + this.kind);
+    }
+  }
+
+  parse() {
+    switch (this.kind) {
+      case ElementKind.Pluck:
+      /* No-op, we just use `raw` */
+        break;
+      case ElementKind.Index:
+      case ElementKind.Pick:
+      case ElementKind.Omit:
+      case ElementKind.Prefix:
+      case ElementKind.Suffix:
+      case ElementKind.Filter:
+        this.brackets = new Brackets(this.raw);
+        break;
+       default:
+         throw new Error('Invalid PathElement kind: ' + this.kind);
+    }
+  }
+
+  selectPluck(target: any): any | undefined {
+    return target[this.raw]
+  }
+
+  selectIndex(target: any): any | undefined {
+    if (!this.brackets) throw Error('Invalid brackets state!' + this);
+
+    let prop = target[this.brackets.prop];
+    if (typeof prop === 'undefined') return undefined;
+    if (typeof prop.length === 'undefined') return undefined;
+
+    let index = this.brackets.op.index;
+    if (index >= prop.length) return undefined;
+    if (index < 0) {
+      index = prop.length + index
+    }
+    if (index < 0) return undefined;
+
+    try {
+      return prop[index];
+    } catch (e) {
+      console.log('Index fail', this.brackets, e);
```

Comment:
In the future we might remove console.logs (might need to get a linter to assist).

Reply:
Yeah, I generally remove them but missed this one.